src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java

```diff
@@ -0,0 +1,3122 @@
+	 *
+	 * @param stats Map of statistics that contains statistics to be removed
+	 * @param controls Set of controls that contains AdvancedControllableProperty to be removed
+	 * @param listKeys list key of statistics to be removed
+	 */
+	private void removeUnusedStatsAndControls(Map<String, String> stats, List<AdvancedControllableProperty> controls, Set<String> listKeys) {
+		for (String key : listKeys) {
+			stats.remove(key);
+			controls.removeIf(advancedControllableProperty -> advancedControllableProperty.getName().equals(key));
+		}
+	}
+
+	//--------------------------------------------------------------------------------------------------------------------------------
+	//endregion
+
+	/**
+	 * calculating minimum of polling interval
+	 *
+	 * @throws ResourceNotReachableException when get limit rate exceed error
+	 */
+	private int calculatingLocalPollingInterval() {
+
+		try {
+			int pollingIntervalValue = SmartThingsConstant.MIN_POLLING_INTERVAL;
+			if (StringUtils.isNotNullOrEmpty(pollingInterval)) {
+				pollingIntervalValue = Integer.parseInt(pollingInterval);
+			}
+
+			int minPollingInterval = calculatingMinPollingInterval();
+			if (pollingIntervalValue < minPollingInterval) {
+				logger.error(String.format("invalid pollingInterval value, pollingInterval must greater than: %s", minPollingInterval));
+				return minPollingInterval;
+			}
+			return pollingIntervalValue;
+		} catch (Exception e) {
+			throw new ResourceNotReachableException(String.format("Unexpected pollingInterval value: %s", pollingInterval));
+		}
+	}
+
+	/**
+	 * calculating minimum of polling interval
+	 *
+	 * @throws ResourceNotReachableException when get limit rate exceed error
+	 */
+	private int calculatingMinPollingInterval() {
+		if (!deviceIds.isEmpty()) {
+			return IntMath.divide(deviceIds.size()
+					, SmartThingsConstant.MAX_THREAD_QUANTITY * SmartThingsConstant.MAX_DEVICE_QUANTITY_PER_THREAD
+					, RoundingMode.CEILING);
+		}
+		return SmartThingsConstant.MIN_POLLING_INTERVAL;
+	}
+
+	/**
+	 * calculating minimum of polling interval
+	 *
+	 * @throws ResourceNotReachableException when get limit rate exceed error
+	 */
+	private int calculatingThreadQuantity() {
+		if (deviceIds.isEmpty()) {
+			return SmartThingsConstant.MIN_THREAD_QUANTITY;
+		}
+		if (deviceIds.size() / localPollingInterval < SmartThingsConstant.MAX_THREAD_QUANTITY * SmartThingsConstant.MAX_DEVICE_QUANTITY_PER_THREAD) {
+			return IntMath.divide(deviceIds.size(), localPollingInterval * SmartThingsConstant.MAX_DEVICE_QUANTITY_PER_THREAD, RoundingMode.CEILING);
+		}
+		return SmartThingsConstant.MAX_THREAD_QUANTITY;
+	}
+
+	/**
+	 * Handle rate limit exceed error while controlling
+	 *
+	 * @param response ResponseEntity
+	 * @throws ResourceNotReachableException when get limit rate exceed error
+	 */
+	private void handleRateLimitExceed(ResponseEntity<?> response) {
+		Optional<String> rateLimit = Optional.ofNullable(response)
```

Comment:
Is it possible to achieve this with the http interceptor? Similarly to the way 429 is handled in XiO aggregator

Reply:
The adapter will reach the rate limit when we send too many control requests in a minute. I think we don't need to build the HTTP interceptor.